File: tf2/tf_python_ws/src/static_broadcaster/static_broadcaster/turtle_tf2_broadcaster.py
# Copyright 2021 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from geometry_msgs.msg import TransformStamped

import numpy as np

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node

from tf2_ros import TransformBroadcaster

from turtlesim.msg import Pose


# This function is a stripped down version of the code in
# https://github.com/matthew-brett/transforms3d/blob/f185e866ecccb66c545559bc9f2e19cb5025e0ab/transforms3d/euler.py
# Besides simplifying it, this version also inverts the order to return x,y,z,w, which is
# the way that ROS prefers it.
# Phép biển đổi roll-pitch-yaw sang q
def quaternion_from_euler(ai, aj, ak):
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk

    q = np.empty((4, ))
    q[0] = cj*sc - sj*cs
    q[1] = cj*ss + sj*cc
    q[2] = cj*cs - sj*sc
    q[3] = cj*cc + sj*ss

    return q


class FramePublisher(Node):

    def __init__(self):
        super().__init__('turtle_tf2_frame_publisher')

        # Declare and acquire `turtlename` parameter
        # Khai báo và truy cập param turtlename, khi khởi chạy thì nó sẽ lấy là turtle1, tên turtle chỉ là gán mặc định cho param
        self.turtlename = self.declare_parameter(
            'turtlename', 'turtle').get_parameter_value().string_value

        # Initialize the transform broadcaster
        # Khởi tạo biến đổi broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)

        # Subscribe to a turtle{1}{2}/pose topic and call handle_turtle_pose
        # callback function on each message
        self.subscription = self.create_subscription(
            Pose, # Loại msg
            f'/{self.turtlename}/pose', # Lấy pose của turtle 1 hoặc 2 
            self.handle_turtle_pose, # Gọi hàm lắng nghe sự thay đổi pose của con rùa
            1)
        self.subscription  # prevent unused variable warning

    def handle_turtle_pose(self, msg):
        t = TransformStamped()

        # Read message content and assign it to
        # corresponding tf variables
        t.header.stamp = self.get_clock().now().to_msg() # Lấy thời gian msg được publish
        t.header.frame_id = 'world' # mặc định con rùa khởi chạy sẽ được sinh trong world
        t.child_frame_id = self.turtlename # lấy id của con rùa

        # Turtle only exists in 2D, thus we get x and y translation
        # coordinates from the message and set the z coordinate to 0
        # Con rùa nằm trong tọa độ là 2D nên chỉ lấy x và y, cho z = 0
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0

        # For the same reason, turtle can only rotate around one axis
        # and this why we set rotation in x and y to 0 and obtain
        # rotation in z axis from the message
        # Con rùa nằm trong tọa độ 2D nên nó có thể thực hiện được duy nhất phép xoay yaw ( theo trục z), 
        # roll (x), pitch(y) sẽ cho bằng 0
        q = quaternion_from_euler(0, 0, msg.theta)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

            # Send the transformation | gửi đi phép biến đổi
        self.tf_broadcaster.sendTransform(t)


def main():
    try:
        rclpy.init()
        node = FramePublisher()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass

File: tf2/tf_cpp_ws/src/learning_tf2_cpp/src/turtle_tf2_broadcaster.cpp
#include <functional>
#include <memory>
#include <sstream>
#include <string>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"
#include "turtlesim/msg/pose.hpp"

class FramePublisher : public rclcpp::Node{
    public:
    FramePublisher(): Node("turtle_tf2_frame_broadcaster"){
        // Lấy tên con rùa
        turtlename_ = this->declare_parameter<std::string>("turtlename","turtle1");
        // Khởi tạo broadcaster
        /*
            Khởi tạo dùng unique đảm bảo:
            + tf2_broadcaster luôn luôn thuộc về một chủ sở hữu là: FramePublisher
            + Không cho phép tham chiếu đến tf2_broadcaster từ bất kì nguồn nào khác
            + Vòng đời rõ ràng, an toàn bộ nhớ vì nó sẽ sống chết với node này.
        */
        tf2_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
        // Subcribe turtle{1|2}
        std::ostringstream stream;
        stream<<"/"<<turtlename_.c_str()<<"/pose";
        std::string topic_name = stream.str();
        // Tạo subcriber lắng nghe pose của con rùa
        subcription_ = this->create_subscription<turtlesim::msg::Pose>(
            topic_name,
            10,
            std::bind(&FramePublisher::handle_turtle_pose,this,std::placeholders:: _1 )
        );


    }

    private:
    void handle_turtle_pose(const std::shared_ptr<const turtlesim::msg::Pose> msg){
        // Tạo transform
        geometry_msgs::msg::TransformStamped t;
        //
        t.header.stamp = this->get_clock()->now();
        t.header.frame_id = "world"; // Cha
        t.child_frame_id = turtlename_.c_str(); // Con
        /*
            Vì con rùa nó chỉ nằm trên mặt phẳng 2D nên:
            + Chỉ đi được theo trục x và y ( z không thể )
            + Chỉ xoay được theo trục z
        */
        t.transform.translation.x = msg->x;
        t.transform.translation.y = msg->y;
        t.transform.translation.z = 0;
        // Chuyển góc quay từ Euler sang Quanternion
        tf2::Quaternion q;
        q.setRPY(0,0,msg->theta);
        t.transform.rotation.x = q.x();
        t.transform.rotation.y = q.y();
        t.transform.rotation.z = q.z();
        t.transform.rotation.w = q.w();

        // Gửi transform đi
        tf2_broadcaster_ ->sendTransform(t);

    }
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subcription_;
    // Dùng unique_ptr vì TransformBroadcaster chỉ có một owner (node này)
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf2_broadcaster_;
    std::string turtlename_;
};

int main(int argc, char **argv){
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<FramePublisher>());
    rclcpp::shutdown();
    return 0;
}